op-chain-ops/script/README.md

```diff
@@ -0,0 +1,139 @@
+# Forge Scripts in Go
+
+This package provides Forge scripting functionality in Go. It allows you to write scripts in Solidity, but call them
+from Go using an in-memory EVM. This is useful for building tooling, since it eliminates the need to shell out to Forge
+or install the Solidity toolchain.
+
+# Usage
+
+To use the Forge scripts, you'll need to create a script host. A minimal example of how to do this below:
+
+```go
+artifactsFS := foundry.OpenArtifactsDir("path/to/forge-artifacts")
+host := NewHost(logger, artifactsFS, nil, DefaultContext, nil)
+```
+
+You can pass in optional arguments to `NewHost` to customize its behavior:
+
+- `WithBroadcastHook` will allow you to broadcast the transactions initiated using `vm.broadcast` on chain via an
+  external broadcaster.
+- `WithIsolatedBroadcast` will clear the warm state between calls, which allows broadcasted gas estimates to be more
+  accurate.
+- `WithCreate2Deployer` will configure the `CREATE2` opcode to broadcast transactions to the Arachnid `Create2Deployer`.
+
+Typically, Forge scripts in Go are implemented using the following pattern:
+
+1. Define a series of simple input and output contracts in Solidity.
+2. Implement your script, and have it take the input/output contracts as arguments to the `run` method.
+3. Define a struct that matches the input/output contracts in Go.
+4. Etch the input and output structs as contracts in Go, then call `run` using the etched input/output contract
+   addresses.
+5. Your script will run, and the output contract will contain the results.
+
+For an example, see [`opsm/implementations.go`](../deployer/opsm/implementations.go).
+
+# Known Limitations
+
+- Structs are not supported in contracts etched from Go code.
+- Bytecode containing linked libraries are not currently supported.
+
+# Best Practices
+
+## Keep I/O Contracts Simple
+
+I/O contracts are the bridge between the Go code and Solidity. You should keep them as simple as possible. Specifically:
+
+1. Use only primitive types in the I/O contracts. This means no structs, arrays, or mappings. The Go script engine will
+   automatically generate getters/setters for public struct fields.
+2. Avoid defining logic in your I/O contracts. Any logic you add in Solidity will have to be reimplemented in Go. If you
+   do add logic to your I/O contract (e.g., for validation) keep it simple.
+3. Avoid mutating your I/O contracts once they have been created.
```

Comment:
what exactly does 'mutating' mean here?

Reply:
Would it be correct to say that the contracts become 'read only', and cannot be written to?